hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -243,7 +255,24 @@ public String getPrimaryGroup() {
 
   @Override
   public void close() throws IOException {
-    IOUtils.cleanupWithLogger(LOG, client);
+    List<ListenableFuture<?>> futures = new ArrayList<>();
+    for (SelfRenewingLease lease : leaseRefs.keySet()) {
+      if (lease == null) {
+        continue;
+      }
+      ListenableFuture<?> future = client.submit(() -> lease.free());
```

Comment:
do we need to worry about running out of workers here, timeouts etc?

Reply:
I am not sure. This is kind of a failsafe; I would not expect code to close the filesystem without closing output streams first, so ideally this loop will only be double checking that free() has already been called on these lease instances without actually performing REST ops. If the output streams opened for this FS instance (for designated single writer directories) have not been closed before the filesystem is closed, this would perform a REST op for each one.